util/overlaydiffer/differ_linux.go

```diff
@@ -1,40 +1,120 @@
 //go:build linux
 // +build linux
 
-package cache
+package overlaydiffer
 
 import (
 	"bytes"
 	"context"
+	"crypto/rand"
+	"encoding/base64"
 	"fmt"
 	"io"
 	"io/ioutil"
 	"os"
 	"path/filepath"
+	"strconv"
 	"strings"
 	"syscall"
+	"time"
 
 	"github.com/containerd/containerd/archive"
 	ctdcompression "github.com/containerd/containerd/archive/compression"
 	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/diff"
 	"github.com/containerd/containerd/errdefs"
 	"github.com/containerd/containerd/mount"
 	"github.com/containerd/continuity/devices"
 	"github.com/containerd/continuity/fs"
 	"github.com/containerd/continuity/sysx"
+	"github.com/klauspost/compress/zstd"
+	"github.com/moby/buildkit/snapshot"
 	digest "github.com/opencontainers/go-digest"
 	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
 	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
 	"golang.org/x/sys/unix"
+
+	log "github.com/moby/buildkit/util/bklog"
 )
 
+func NewOverlayDiffer(store content.Store, sn snapshot.Snapshotter, d diff.Comparer) diff.Comparer {
+	return &overlayDiffer{
+		store: store,
+		d:     d,
+		sn:    sn,
+	}
+}
+
+const containerdUncompressed = "containerd.io/uncompressed"
+
 var emptyDesc = ocispecs.Descriptor{}
 
-// computeOverlayBlob provides overlayfs-specialized method to compute
-// diff between lower and upper snapshot. If the passed mounts cannot
-// be computed (e.g. because the mounts aren't overlayfs), it returns
-// an error.
-func (sr *immutableRef) tryComputeOverlayBlob(ctx context.Context, lower, upper []mount.Mount, mediaType string, ref string, compressorFunc compressor) (_ ocispecs.Descriptor, ok bool, err error) {
+// overlayDiffer provides overlayfs-specialized method to compute
+// diff between lower and upper snapshot.
+type overlayDiffer struct {
+	store content.Store
+	d     diff.Comparer
+	sn    snapshot.Snapshotter
+}
+
+// Compare creates a diff between the given mounts and uploads the result
+// to the content store.
+func (d *overlayDiffer) Compare(ctx context.Context, lower, upper []mount.Mount, opts ...diff.Opt) (desc ocispecs.Descriptor, err error) {
+
+	// Determine differ and error/log handling based on envvar and snapshotter.
+	var enableOverlay, fallback, logWarnOnErr bool
+	if forceOvlStr := os.Getenv("BUILDKIT_DEBUG_FORCE_OVERLAY_DIFF"); forceOvlStr != "" {
+		enableOverlay, err = strconv.ParseBool(forceOvlStr)
+		if err != nil {
+			return emptyDesc, errors.Wrapf(err, "invalid boolean in BUILDKIT_DEBUG_FORCE_OVERLAY_DIFF")
+		}
+		fallback = false // prohibit fallback on debug
+	} else {
+		enableOverlay, fallback = true, true
+		switch d.sn.Name() {
+		case "overlayfs", "fuse-overlayfs", "stargz":
+			logWarnOnErr = true // snapshotter should support overlay diff. so print warn log on failure
```

Comment:
Does “fuse-overlayfs” really support overlay diff?

Reply:
Thank you for pointing this. No, it doesn't because fuse-overlayfs snapshotter doesn't provide overlayfs mounts. Disabled it for fuse-overlayfs snapshotter.